source/io/module_javaScriptInvoke.js

```diff
@@ -358,30 +372,52 @@
                 newErrorCode = ERRORS.kNIUnableToFindFunctionForJavaScriptInvoke.CODE;
                 newErrorSource = ERRORS.kNIUnableToFindFunctionForJavaScriptInvoke.MESSAGE + '\'' + functionName + '\'.';
                 Module.coreHelpers.mergeErrors(newErrorStatus, newErrorCode, newErrorSource, errorStatusPointer, errorCodePointer, errorSourcePointer);
+                Module.eggShell.setOccurrenceAsync(occurrencePointer);
                 return;
             }
 
-            var context = undefined;
+            var returnTypeName = getParameterTypeString(returnPointer, 0);
+            var returnValuePointer = undefined;
+            if (returnTypeName !== 'StaticTypeAndData') {
+                // User doesn't want return value. We're passing '*' for the return in VIA code, we get StaticTypeAndData
+                returnValuePointer = JavaScriptInvoke_GetParameterPointer(returnPointer, 0);
+            }
+            if (returnTypeName === 'Array') {
+                returnTypeName += getArrayElementTypeString(returnValuePointer);
+            }
+
+            var asyncFlag = false;
+            var generateContext = function () {
```

Comment:
I think it would be cleaner to move the generateContext function out next to the generateCompletionCallback function. With the state object discussed below that would be possible

Reply:
Implemented the state object and moved the generateContext function closer to the generateCompletionCallback function.